here-naksha-app-service/src/test/java/com/here/naksha/app/service/WriteFeaturesAtomicityTest.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.app.service;
+
+import static com.here.naksha.app.common.CommonApiTestSetup.setupSpaceAndRelatedResources;
+import static com.here.naksha.app.common.ResponseAssertions.assertThat;
+import static com.here.naksha.app.common.TestUtil.loadFileOrFail;
+
+import com.here.naksha.app.common.ApiTest;
+import com.here.naksha.app.common.NakshaTestWebClient;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+class WriteFeaturesAtomicityTest extends ApiTest {
+
+  private static final NakshaTestWebClient nakshaClient = new NakshaTestWebClient();
+  private static final String SPACE_ID = "write_features_atomicity_test_space";
+
+  public WriteFeaturesAtomicityTest() {
+    super(nakshaClient);
+  }
+
+  @BeforeAll
+  static void prepareEnv() {
+    setupSpaceAndRelatedResources(nakshaClient, "WriteFeaturesAtomicity/setup");
+  }
+
+  @Test
+  void tc_1101_partialCreateShouldFail() throws URISyntaxException, IOException, InterruptedException {
+    // Given: multiple features to save, some of which are invalid
+    List<String> validFeatureIds = List.of("tc_1101_feature_1", "tc_1101_feature_2");
+    String createFeaturesJson = loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/create_features.json");
+
+    // When: saving these features
+    String streamId = UUID.randomUUID().toString();
+    HttpResponse<String> saveResp = nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeaturesJson, streamId);
+
+    // Then: response indicates failure
+    assertThat(saveResp)
+        .hasStatus(500)
+        .hasStreamIdHeader(streamId);
+
+    // And: none of the features got saved
+    String idsQuery = "?id=" + String.join("&id=", validFeatureIds);
+    HttpResponse<String> getResp = nakshaClient.get("hub/spaces/" + SPACE_ID + "/features" + idsQuery, streamId);
+    assertThat(getResp)
+        .hasStatus(200)
+        .hasStreamIdHeader(streamId)
+        .hasJsonBody(loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/empty_response.json"));
+  }
+
+  @Test
+  void tc_1102_partialUpdateShouldFail() throws URISyntaxException, IOException, InterruptedException {
```

Comment:
For Update, I have same comment as create. This doesn't seem to really testing the atomicity (transactional) part. Invalid request may get rejected before psql transaction can happen.
For Update, we can consider a request that has 2 features, with first feature not having any ID / UUID conflict but second feature has UUID conflict with existing feature in storage. And as a result, we should rollback entire transaction, verifying nothing gets saved in storage.

Reply:
done